thrust/benchmarks/bench/transform/basic.cu

```diff
@@ -121,9 +124,10 @@ static void mul(nvbench::state& state, nvbench::type_list<T>)
 
   state.exec(nvbench::exec_tag::no_batch | nvbench::exec_tag::sync, [&](nvbench::launch&) {
     const T scalar = startScalar;
-    thrust::transform(c.begin(), c.end(), b.begin(), [=] __device__ __host__(const T& ci) {
-      return ci * scalar;
-    });
+    thrust::transform(
+      c.begin(), c.end(), b.begin(), cuda::std::allow_copied_arguments([=] __device__ __host__(const T& ci) {
+        return ci * scalar;
+      }));
```

Comment:
This is what uses are expected to write. They need to wrap their callables in `cuda::std::allow_copied_arguments(...)` or specialize `__allows_copied_arguments_impl`.

Reply:
Can you elaborate more on under what conditions `cuda::std::allow_copied_arguments` will be required? Is it for peak performance?

Reply:
The sad answer for now is that you need to wrap your callable in `allow_copied_arguments` every time you call a `thrust` API that may use `thrust::transform` underneath and your callable does not rely on the address of your arguments. We have some ideas on how to detect this automatically for a set of known types and functors, e.g. `thrust::plus<int>`, so this will automatically enable at least some use cases.

Reply:
Is it due to the use of shared memory? If you take address, as it won't be same as the original data?

Reply:
Exactly. `cub::DeviceTranform` on Hopper+ will use an asynchronous bulk copy from global to shared memory and then run the transformation function on the data in shared memory. This generally leads to higher saturation of the memory bandwidth. But if your callable has a `const T& value` parameter that reference will point to shared memory now instead of global memory, so `&value` may not be what you expect.

Reply:
What if I use `const T value`?

Reply:
Like @mfbalin, I thought we were going to be able to detect when the user provided callable is passing arguments by reference or by value and implicitly enable these optimizations when everything is passed by value.

Reply:
> What if I use `const T value` as the callable argument? Then are the optimizations enabled by default?

This is what I plan to do after this PR. I will try further detections for function objects we know, like `thrust::plus` or `cuda::std::plus`. Notice though, that `const T value` does not work for any `T`. If your `T` is `thrust::device_reference<U>`, or `std::reference_wrapper<U>` or any other kind of language-wise value-type with referential semantic (e.g. all proxy reference types), we cannot opt into the fast path.

@jrhemstad we aim to get as far as possible with the detection :)

Reply:
I just want to understand the limitations deeply, sorry if I am missing something with my limited experience. If `T` is a reference, then can't the address of the reference be copied using the fast path as well, in that case, I would assume the address is dependent on the index.

Reply:
When the user uses as `const T value` as the callable argument, the address can not be taken anymore. So can the fast past optimization be enabled when the callable argument is passed by value?

Reply:
If `const T value` happens to be `const std::reference_wrapper<U> value`, or `const thrust::device_reference<U> value`, or any other proxy reference type, and your buffer contains `U`s, you can still recover the address of the `U`s in global memory even though your function syntactically takes arguments by copy.

Reply:
So you want to fix it in such a way that even for those types, the fast path can be used to dereference the reference_wrapper object?

Reply:
I want to understand if there is any disadvantage if the user has a kernel that takes `const T value` and the user always wraps it with `cuda::std::allow_copied_arguments` vs not wrapping with it.